src/main/java/io/confluent/connect/hdfs/wal/FSWAL.java

```diff
@@ -183,13 +189,141 @@ private void commitEntriesToStorage(Map<WALEntry, WALEntry> entries) {
     }
   }
 
+  /**
+   * Extract the latest offset and file path from the WAL file.
+   * Attempt with the most recent WAL file, and fall back to the old file if it's not applicable.
+   *
+   * <p> The old WAL is used when the most recent WAL file has already been truncated
+   * and is not existent, which may happen when the connector has flushed all records,
+   * and is restarted. </p>
+   *
+   * @return the latest offset and filepath from the WAL file or null
+   */
+  public Pair<Long, String> extractLatestOffsetFromWAL() {
+    String oldWALFile = logFile + OLD_LOG_EXTENSION;
+    try {
+      Pair<Long, String> latestOffset = null;
+      if (storage.exists(logFile)) {
+        log.trace("Restoring offset from WAL file: {}", logFile);
+        if (reader == null) {
+          reader = newWalFileReader();
+        } else {
+          // reset read position after apply()
+          reader.seekToFirstRecord();
+        }
+        List<String> committedFileBatch = getLastFilledBlockFromWAL(reader);
+        // At this point the committedFilenames list will contain the
+        // filenames in the last BEGIN-END block of the file. Find the latest offsets among these.
+        latestOffset = getLatestOffsetFromList(committedFileBatch);
+      }
+
+      // attempt to use old log file if recent WAL is empty or non-existent
+      if (latestOffset == null && storage.exists(oldWALFile)) {
+        log.trace("Could not find offset in log file {}. Using {} instead", logFile, oldWALFile);
+        Reader oldFileReader =
+            new WALFile.Reader(conf.getHadoopConfiguration(), Reader.file(new Path(oldWALFile)));
+        List<String> committedFileBatch = getLastFilledBlockFromWAL(oldFileReader);
+        latestOffset = getLatestOffsetFromList(committedFileBatch);;
+        oldFileReader.close();
+      }
+      return latestOffset;
+    } catch (IOException e) {
+      log.warn("Error restoring offsets from either {} or {} WAL files", logFile, oldWALFile);
+      return null;
+    }
+  }
+
+  /**
+   * Extract the last filled BEGIN-END block of entries from the WAL.
```

Comment:
This would also return entries after the last BEGIN with no END, is that a concern? Should we throw an exception in that case?

Reply:
When we write a block of entries in [appendToWal() here](https://github.com/confluentinc/kafka-connect-hdfs/blob/c5073c7f831e377dbd1f63b9b08ccd40d6dcfb92/src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java#L813) we always add the END right after. To be safe, we can definitely throw an exception.

Reply:
Thanks, I saw that, but I was thinking of something happening to that call before writing the end marker. Not sure if we catch that scenario earlier in the code. If there's not an obvious place where we're doing that, I would throw here.

Reply:
Thanks @erdody it is a good point, I revised the logic to handle the missing end marker accordingly. Take a look if you get a chance.